R/BackendHdf5-class.R

```diff
@@ -0,0 +1,149 @@
+    msg <- c(.valid.BackendHdf5.md5sum(object@files, object@md5sum),
+             .valid.BackendHdf5.hdf5file(object@hdf5file, object@files),
+             .valid.BackendHdf5.hdf5file.exist(object@hdf5file))
+    if (length(msg)) msg
+    else TRUE
+})
+
+#' Initialize the Hdf5 backend, i.e. create the hdf5 files and add their names
+#' to the @hdf5file slot.
+#'
+#' @param object `BackendHdf5`
+#'
+#' @param files `character` with the original file names.
+#'
+#' @param spectraData `DataFrame` with the spectrum metadata.
+#'
+#' @param path `character(1)` defining the path where the hdf5 files should be
+#'     stored.
+#'
+#' @return `BackendHdf5` object.
+#'
+#' @author Johannes Rainer
+#'
+#' @md
+#'
+#' @noRd
+setMethod("backendInitialize", "BackendHdf5", function(object, files,
+                                                       spectraData, path = ".",
+                                                       ...) {
+    path <- suppressWarnings(normalizePath(path))
+    dir.create(path, showWarnings = FALSE, recursive = TRUE)
+    n_files <- length(files)
+    object@files <- files
+    object@md5sum <- character(n_files)
+    object@hdf5file <- character(n_files)
+    comp_level <- .hdf5_compression_level()
+    for (i in seq_len(n_files)) {
+        h5file <- paste0(path, "/", digest(files[i]), ".h5")
+        if (file.exists(h5file))
+            stop("File ", h5file, " does already exist. Please use a different path.")
+        h5 <- H5Fcreate(h5file)
+        h5createGroup(h5, "spectra")
+        h5createGroup(h5, "md5")
+        H5Fclose(h5)
+        object@hdf5file[i] <- h5file
+    }
+    validObject(object)
+    object
+})
+
+#' Import the data from the raw MS files and store them to the hdf5 files.
+#'
+#' @inheritParams backendInitialize
+#'
+#' @return `BackendHdf5`
+#'
+#' @author Johannes Rainer
+#'
+#' @md
+#'
+#' @noRd
+setMethod("backendImportData", "BackendHdf5", function(object, spectraData,
+                                                       ...,
+                                                       BPPARAM = bpparam()) {
+    object@md5sum <- bpmapply(fileNames(object), object@hdf5file,
+                              FUN = .serialize_msfile_to_hdf5,
+                              BPPARAM = BPPARAM)
+    validObject(object)
+    object
+})
+
+## setMethod("backendReadSpectra", "BackendHdf5", function(object, spectraData,
+##                                                         ...) {
+## })
+
+## setMethod("backendWriteSpectra", "BackendHdf5", function(object, spectra,
+##                                                          spectraData) {
+## })
+
+#' Write the content of a single mzML/etc file to an h5file. We're using the
+#' spectrum index in the file as data set ID.
+#'
+#' @return `character(1)` with the md5 sum of the spectra data.
+#'
+#' @author Johannes Rainer
+#'
+#' @noRd
+.serialize_msfile_to_hdf5 <- function(file, h5file) {
+    h5 <- H5Fopen(h5file)
```

Comment:
I like the `on.exit(H5close(h5))` approach because it closes the file even if we have an error before the function is correctly finished.